Ensure that file:write_file/3 can use `writev` if given `raw` mode

nathanl opened the pull request:
Previously, this function would turn any input into a single binary
before writing. This meant it could not take advantage of the `writev`
system call if it was given a list of binaries and told to write with
`raw` mode.

To see this, start an erlang shell on the parent commit, and also
start [this dtrace script](https://github.com/evanmiller/tracewrite) like this:

    sudo dtrace -s tracewrite.d -p $(pgrep beam)

In the erlang shell, run the following:

```erlang
file:write_file("/tmp/tmp.txt", [<<97,98>>, <<98,97>>], [raw]).
```

In the dtrace output, you will see that the system call used is `write`.

Now repeat with this commit, and you will see that `writev` is used.

This is a very small change, but looks bigger because of whitespace
differences. See https://github.com/erlang/otp/pull/1149/files?w=1 to
ignore them.

nathanl commented:
I originally tried to work around this issue in Elixir, but @josevalim and @fishcakez suggested (wisely) that it was best to fix it here. See https://github.com/elixir-lang/elixir/pull/5154 for our discussion.

lexmag commented on lib/kernel/src/file.erl:
```diff
@@ -398,24 +398,32 @@ write_file(Name, Bin) ->
+        end
+  end.
+
+do_write_file(Name, B, ModeList) ->
```
Maybe we can use `Bin` variable name instead of single-letter `B`, it better communicates the meaning.

nathanl replied:
Yeah, I was unsure here. The input could actually be a list of binaries or a single binary. Would `Data` be good here, and maybe in the signature of `write_file` too?

lexmag replied:
👍 for `Data`, reads much better than inexpressive `B`.

fenollp replied:
Neither are expressive IMO. I think adding a spec for `do_write_file/3` would help more.
How about both?

nathanl replied:
@lexmag I renamed some variables - see what you think. (I went with `IOData` instead of `Data`.)

@fenollp Good idea about adding a spec. Can you tell me where to do that?

nathanl replied:
@fenollp Actually, on second thought, I'm not whether a spec here makes sense. `do_write_file/3` is a private implementation detail of `write_file/3`; if it seems better, it could be refactored to be part of `write_file/3` with no effect on callers. As such, I think it should be covered by any tests of `write_file/3`.

If additional testing seems good, we might test the following: "if given an iolist and a mode list including `raw`, `write_file/3` will call `write` with its iolist, not with a binary". But I don't know how to write that test. I'm not even sure where the test would go - I'm guessing `read_write_file/1` in `lib/kernel/test/file_SUITE.erl`, but then it appears to me that the existing tests there are using a `write_file/2` that's defined at the bottom of that file, not the one touched by this PR.

OTP-Maintainer commented:
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: Ensure that file:write_file/3 can use `writev` if given `raw` mode


----------------

I am a script, I am not human 

----------------

nathanl commented:
> The summary line of the commit message is too long

If we get this PR to a point where everyone is happy with the changes, I'll squash it back to one commit and shorten the summary line.

josevalim commented:
> Actually, on second thought, I'm not sure whether a spec here makes sense.

I agree. We can write specs for private functions but I don't think there is much to gain here given we already have a public spec and the fact the function is quite small.

I am not sure we could write a test as well. FWIW, this patch looks good to me. 👍

bjorng commented:
Looks good, but I have a few nits to pick.

Since this is an optimization and not a bug fix, it would be better to include it in master instead of maint.

This branch changes the error behavior:

```
2> file:write_file("xxx", [a], [raw]).
{error,badarg}
3> 
=ERROR REPORT==== 31-Aug-2016::15:20:22 ===
Bad value on output port 'efile'

```
(The error report is new.) The iodata could be checked using `iolist_size/1`. For the typical use case when the iodata is a binary or a list of binaries, that should be very fast.

We will also need test cases. (There does not seem to be any for `file:write_file/3`, for some reason.) Changing code without test cases make us nervous.

`read_write_file/1` in `file_SUITE` could be extended. I suggest that you introduce a helper function and change `read_write_file/1` to use it:

```
do_read_write_file(Name, Data) ->
    BinData = iolist_to_binary(Data),

    Res = ?FILE_MODULE:write_file(Name, Data),
    {ok,BinData} = ?FILE_MODULE:read_file(Name),

    Res = ?FILE_MODULE:write_file(Name, Data, []),
    {ok,BinData} = ?FILE_MODULE:read_file(Name),

    Res = ?FILE_MODULE:write_file(Name, Data, [raw]),
    {ok,BinData} = ?FILE_MODULE:read_file(Name),
    
    BinData.

```


This helper function will only handle the success cases. It is probably easiest to make another helper function for the error cases.

OTP-Maintainer commented:
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: Ensure that file:write_file/3 can use `writev` if given `raw` mode


----------------

I am a script, I am not human 

----------------

OTP-Maintainer commented:
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: Ensure that file:write_file/3 can use `writev` if given `raw` mode


----------------

I am a script, I am not human 

----------------

nathanl commented:
@bjorng 

> Since this is an optimization and not a bug fix, it would be better to include it in master instead of maint.

I'm planning to rebase all this into logical commits once it looks good. Maybe that would be a good time to rebase on master and point the PR there?

> This branch changes the error behavior

I've fixed that now - I think what I did is what you were suggesting. `iolist_size/1` raises if the iolist contains an atom, and I catch that.

> We will also need test cases. (There does not seem to be any for file:write_file/3, for some reason.) Changing code without test cases make us nervous.

Totally agree. Being a newbie to Erlang, I was unsure how to work with the test suite, so I'd been doing manual testing in `erl`. 😝  I wasn't quite clear on what you meant by the helper function, but I think the tests I've added accomplish the same thing. Do they look OK? Is this enough testing?

OTP-Maintainer commented:
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: Ensure that file:write_file/3 can use `writev` if given `raw` mode


----------------

I am a script, I am not human 

----------------

lexmag commented on lib/kernel/src/file.erl:
```diff
@@ -397,25 +397,38 @@ write_file(Name, Bin) ->
+        true ->
+          %% For backwards compatibility of error messages
+          try iolist_size(IOData) of
+            _size -> do_write_file(Name, IOData, ModeList)
```
`_size` –> `_Size` (or `_`).

lexmag commented on lib/kernel/test/file_SUITE.erl:
```diff
@@ -535,6 +535,20 @@ read_write_file(Config) when is_list(Config) ->
     {Bin1,Bin2} = split_binary(Bin5,byte_size(Bin1)),
 
+    %% Try writing an iolist
+    IoList = [<<97,98>>, <<98,97>>],
```
Shouldn't it be named `IOList`?

lexmag commented on lib/kernel/test/file_SUITE.erl:
```diff
@@ -535,6 +535,20 @@ read_write_file(Config) when is_list(Config) ->
 
+    %% Try writing an iolist
+    IoList = [<<97,98>>, <<98,97>>],
+    IoAsBinary = erlang:iolist_to_binary(IoList),
```
`IoAsBinary` looks a bit confusing.

lexmag commented on lib/kernel/test/file_SUITE.erl:
```diff
@@ -535,6 +535,20 @@ read_write_file(Config) when is_list(Config) ->
+    IoList = [<<97,98>>, <<98,97>>],
+    IoAsBinary = erlang:iolist_to_binary(IoList),
+    ok = ?FILE_MODULE:write_file(Name, IoList),
+    {ok,IoAsBinary} = ?FILE_MODULE:read_file(Name),
```
Space after `,`?

nathanl commented:
@lexmag I pushed a commit based on your suggestions.

OTP-Maintainer commented:
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: Ensure that file:write_file/3 can use `writev` if given `raw` mode


----------------

I am a script, I am not human 

----------------

nathanl commented:
@bjorng I rebased this PR on master and re-pointed it at master. I also made it two commits: one to add tests for `file:write_file/3`, and a second for the optimization.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

bjorng commented on lib/kernel/src/file.erl:
```diff
@@ -397,25 +397,38 @@ write_file(Name, Bin) ->
+    end.
+
+do_write_file(Name, IOData, ModeList) ->
+    case open(Name, [binary, write |
```
The calls to `lists:delete/2` are not necessary; duplicated options will be ignored.

I know that the original code does the same thing, but there is no reason to keep the duplication in your refactored code.

nathanl replied:
Great! Glad to delete it. Thanks.

nathanl commented:
@bjorng Great, thanks! I've rebased on master as you suggested.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

nathanl commented:
Yay! Thanks @bjorng for giving feedback and merging this!

Also thanks @iamvery and @michaelgundlach for looking at this code with me!
